File: 494-target-sum/target-sum.cpp
class Solution {
public:
    unordered_map<string, int> cache;
    int dfs(int index, int sum, vector<int>& nums, int target){
        if (index == nums.size()) return sum == target ? 1 : 0;
        string key = to_string(index) + "_" + to_string(sum);
        if (cache.contains(key)) return cache[key];
        cache[key] = dfs(index + 1, sum + nums[index], nums, target) + dfs(index + 1, sum - nums[index], nums, target);
        return cache[key];
    }

    int findTargetSumWays(vector<int>& nums, int target) {
        return dfs(0, 0, nums, target);
    }
};

File: 494-target-sum/target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        cache = {}
        def dfs(index, total):
            if index == len(nums): return 1 if total == target else 0
            if (index, total) in cache: return cache[(index, total)]

            cache[(index, total)] = dfs(index + 1, total + nums[index]) + dfs(index + 1, total - nums[index])
            return cache[(index, total)]
        return dfs(0,0)